backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+              ExecutionCtxs.contextInstance().getHandle,
+              Array(ColumnarBatches.getNativeHandle(batch)),
+              NativeMemoryManagers
+                .contextInstance("ColumnarCachedBatchSerializer")
+                .getNativeInstanceHandle
+            )
+            CachedColumnarBatch(
+              Math.toIntExact(serializeResult.getNumRows),
+              serializeResult.getSerialized.length,
+              serializeResult.getSerialized)
+          }
+        }
+    }
+  }
+
+  override def convertCachedBatchToColumnarBatch(
+      input: RDD[CachedBatch],
+      cacheAttributes: Seq[Attribute],
+      selectedAttributes: Seq[Attribute],
+      conf: SQLConf): RDD[ColumnarBatch] = {
+    val execHandle = ExecutionCtxs.contextInstance().getHandle
+    // Find the ordinals and data types of the requested columns.
+    val requestedColumnIndices = selectedAttributes.map {
+      a => cacheAttributes.map(_.exprId).indexOf(a.exprId)
+    }
+    val arrowAlloc = ArrowBufferAllocators.contextInstance()
+    val timezoneId = SQLConf.get.sessionLocalTimeZone
+    val localSchema = toStructType(selectedAttributes)
+    val arrowSchema = SparkArrowUtil.toArrowSchema(localSchema, timezoneId)
+    val cSchema = ArrowSchema.allocateNew(arrowAlloc)
+    ArrowAbiUtil.exportSchema(arrowAlloc, arrowSchema, cSchema)
+    val deserializer = {
+      val handle = ColumnarBatchSerializerJniWrapper.INSTANCE.init(
+        cSchema.memoryAddress(),
+        execHandle,
+        NativeMemoryManagers
+          .contextInstance("ColumnarCachedBatchSerializer")
+          .getNativeInstanceHandle
+      )
+      cSchema.close()
+      TaskResources.addRecycler(
+        s"ColumnarCachedBatchSerializer_convertCachedBatchToColumnarBatch_$handle",
+        50) {
+        ColumnarBatchSerializerJniWrapper.INSTANCE.close(
+          execHandle,
+          handle)
+      }
+      handle
+    }
+
+    input.mapPartitions {
+      it =>
+        new CloseableColumnBatchIterator(new Iterator[ColumnarBatch] {
+          override def hasNext: Boolean = {
+            it.hasNext
+          }
+
+          override def next(): ColumnarBatch = {
+            val cachedBatch = it.next().asInstanceOf[CachedColumnarBatch]
+            val batchHandle = ColumnarBatchSerializerJniWrapper.INSTANCE.deserialize(
+              execHandle,
+              deserializer,
+              cachedBatch.encoded)
+            val batch = ColumnarBatches.create(batchHandle, batchHandle)
+
+            val heavy = ColumnarBatches.ensureLoaded(
+              arrowAlloc,
+              batch
+            )
+            val projected = new ColumnarBatch(requestedColumnIndices.map {
+              ordinal =>
+                heavy.column(ordinal)
+            }.toArray, heavy.numRows())
+            ColumnarBatches.retain(projected)
+            ColumnarBatches.release(heavy)
+
+            ColumnarBatches.ensureOffloaded(
```

Comment:
@zhztheplayer  The one more load and offload operator seems worse than before...  we should add a new jni method, something like 
```
deserialize(
    long execHandle,
    long deserializer,
    byte[] cachedBatch.encoded,
    long selectedCSchema
   )
```

it is used to deserialize the columnar batch and only return a offload columnar batch with the selected attribute.

Reply:
BTW, we should return a velox columnar batch but `ensureOffloaded` returns a arrow based.

Reply:
> @zhztheplayer The one more load and offload operator seems worse than before... we should add a new jni method, something like
> 
> ```
> deserialize(
>     long execHandle,
>     long deserializer,
>     byte[] cachedBatch.encoded,
>     long selectedCSchema
>    )
> ```
> 
> it is used to deserialize the columnar batch and only return a offload columnar batch with the selected attribute.

If we do this, we can implement filter-on-read to avoid unnecessary deserializations on the non-requested columns.

It it's tricky more or less, for the first version I recommend adding an abstract function to C++ `ColumnarBatch` API to do a simple ordinal-based projection which is very reasonable to any in memory columnar format. This way we can avoid the Arrow conversion either.